Add support for SwiftPM & Accio

Jeehut opened the pull request:
This adds support for SwiftPM manifest based dependency managers. Specifically this adds support for installing via [Accio](https://github.com/JamitLabs/Accio) but may also work with SwiftPM once it's integrated into Xcode.

Please note that this project is part of Accio's official [integration tests](https://github.com/JamitLabs/Accio/tree/work/1000-frameworks/Demo/Shared/AppDependencies) within the [Demo project](https://github.com/JamitLabs/Accio/tree/work/1000-frameworks/Demo).

SlashDevSlashGnoll commented:
This becomes much more important in XCode 11 as Apple has added direct support for SPM in all xcode projects.  This will soon become the preferred way to manage dependencies.  Getting this in will be very helpful.

matej commented:
Does `Package.swift` apply to XCode 11 as is, or would it require any changes? Anybody tried this?

mergesort commented:
For what it's worth I tested this, and there seems to be a conflict between the module name and class being `MBProgressHUD`.

This is what happens when I try to extend it.
<img width="572" alt="image" src="https://user-images.githubusercontent.com/716513/62016925-a4c8d200-b182-11e9-8973-c5c1b0272e12.png">

And issues at a call-site.
<img width="495" alt="image" src="https://user-images.githubusercontent.com/716513/62016939-b3af8480-b182-11e9-9ae9-cb8acd91abb9.png">

Dschee commented:
@mergesort Could this be related to [that](https://bugs.swift.org/browse/SR-898) open Swift bug?

mergesort commented:
It quite likely could be. I'm wondering what a good fix would be that adds SPM support. All changes seem like they would be breaking, but I assume the easiest would be to change the module name rather than all of the classes?

Dschee commented:
I agree, changing the module name within the `Package.swift` file would probably be the least affecting method of getting it working with SwiftPM. We would probably need to specify the path for the target explicitly and document that the `import` statement for SwiftPM users would be different, maybe just `MBProgress`?

mergesort commented:
That sounds like a good suggestion to me, and after a few days I can't really think of a better way to resolve the conflict.

Schaefers commented:
Can you perform these changes on your fork? Would be happy to see this merge or use your fork.

4brunu commented:
Hey, for those interested, I just created https://github.com/jdg/MBProgressHUD/pull/592 that doesn't have the conflict between the module name and class being `MBProgressHUD`.
If @Jeehut wants to incorporate the changes in this PR or just merge the other PR, I'm fine with it both options, I just want that MBProgressHUD support SPM 🙂

Jeehut commented:
Thank you @4brunu, I just rebased my fork, cherry-picked your changes in a single commit and added a badge that MBProgressHUD supports SwiftPM to the README. I think you can close #592 now ad a duplicate to this. Great work with the modulemap etc.!

4brunu commented:
@Jeehut no problem 🙂 
What do you think about adding the SPM instructions to the readme? 
I saw the Accio instructions but not the SPM.
I know they are almost the same, but it's just to clarify to the users when they look for SPM in the readme.

Jeehut commented:
@4brunu Makes sense. It's just the same, only the command is different, fixed it! 👍

4brunu commented:
Thanks 👍 I have tested this PR locally and it's working, for me it's ready to merge 🙂

matej commented on include/MBProgressHUD/MBProgressHUD.h:
```diff
@@ -0,0 +1 @@
+../../MBProgressHUD.h
```
Could you perhaps elaborate on the significance of this file?  Adding some comments seems appropriate.

4brunu replied:
This file is a symbolic link to satisfy the module.modulemap, sorry but I'm not an expert on why, I used that on some others SPM libraries, I don't remember where did I get this.

4brunu replied:
Sorry about not giving to much technical details on this.
I have used this also in other like https://github.com/zxingify/zxingify-objc/pull/489

4brunu replied:
I do remember now where I got this.
I saw this in the real-cocoa repository.
Here is the commit where they add support for SPM https://github.com/realm/realm-cocoa/commit/26e2704278863baf8a0bef0f8cce10ed0ee3d9cd

This is how it works, first we need to create a directory `include`.
Inside the `include` directory, we create a `module.modulemap` and a subdirectory with the library name, in this case `MBProgressHUD`.
The `module.modulemap` will point at the subdirectory with the library name to search for the headers.
Inside subdirectory with the library name, `MBProgressHUD`, we need to create symbolic links to the headers used by the library.
And in the `Package.swift` in the `publicHeadersPath` we point to the `include` directory.
This way the SPM imports correctly the headers.

In the realm commit, here is the [Package.swift](https://github.com/realm/realm-cocoa/blob/963dbe558b802e3b9b78b810dcbceea3b4853ec9/Package.swift), the [include](https://github.com/realm/realm-cocoa/blob/963dbe558b802e3b9b78b810dcbceea3b4853ec9/include) directory with the [module.modulemap](https://github.com/realm/realm-cocoa/blob/963dbe558b802e3b9b78b810dcbceea3b4853ec9/include/module.modulemap) and the project name sub directory, [Realm](https://github.com/realm/realm-cocoa/tree/963dbe558b802e3b9b78b810dcbceea3b4853ec9/include/Realm) in their case, with all the symbolic links to the headers.

Hope it helps to clarify how this works.

matej commented:
Alright, let's give this a try. I don't use Swift package manager, so I'll just rely on your expertise here. Thank you for your contribution!

4brunu commented:
Nice, thanks 👍 
Could you please also make a new release?
Because it would be nicer to point SPM to a tag instead of a branch or a commit 🙂
